src/js/program/media-controller.js

```diff
@@ -79,30 +87,84 @@ export default class MediaController {
 
     attach() {
         const { model, provider } = this;
-        this.attached = true;
-        model.set('attached', true);
 
-        provider.off('all', model.videoEventHandler, model);
-        provider.on('all', model.videoEventHandler, model);
-
-        if (model.checkComplete()) {
-            model.playbackComplete();
+        if (this.beforeComplete) {
+            this._playbackComplete();
         }
 
-        provider.attachMedia();
-
         // Restore the playback rate to the provider in case it changed while detached and we reused a video tag.
         model.setPlaybackRate(model.get('defaultPlaybackRate'));
+
+        addProviderListeners(this);
+        provider.attachMedia();
+        this.attached = true;
+        model.set('attached', true);
     }
 
     detach() {
         const { model, provider } = this;
+        this.thenPlayPromise.cancel();
+        removeProviderListeners(this);
+        provider.detachMedia();
         this.attached = false;
         model.set('attached', false);
+    }
 
-        model.setThenPlayPromise(cancelable(() => {}));
-        provider.off('all', model.videoEventHandler, model);
-        provider.detachMedia();
+    // Executes the playPromise
+    _playAttempt(playPromise, playReason, item) {
+        const { mediaModel, model, provider } = this;
+
+        this.trigger(MEDIA_PLAY_ATTEMPT, {
+            item,
+            playReason
+        });
+        // Immediately set player state to buffering if these conditions are met
+        if (provider && provider.video && !provider.video.paused) {
+            model.set(PLAYER_STATE, STATE_BUFFERING);
+        }
+
+        playPromise.then(() => {
+            if (!mediaModel.get('setup')) {
+                // Exit if model state was reset
+                return;
+            }
+            mediaModel.set('started', true);
+            if (mediaModel === model.mediaModel) {
+                syncPlayerWithMediaModel(mediaModel);
+            }
+        }).catch(error => {
+            model.set('playRejected', true);
+            const videoTagPaused = provider && provider.video && provider.video.paused;
+            if (videoTagPaused) {
+                mediaModel.set(PLAYER_STATE, STATE_PAUSED);
+            }
+            this.trigger(MEDIA_PLAY_ATTEMPT_FAILED, {
+                error,
+                item,
+                playReason
+            });
+        });
+    }
+
+    _playbackComplete() {
+        const { provider } = this;
+        this.beforeComplete = false;
+        provider.setState(STATE_COMPLETE);
```

Comment:
Where's the attached check? That's what allows postrolls to play without complete firing until after the postroll is done.